File: MicType/Sources/MicType/HistoryStore.swift
import Foundation

struct HistoryItem {
    let date: Date
    let raw: String
    let polished: String
}

/// 最近的听写记录（保留 20 条，持久化到 UserDefaults）
final class HistoryStore {
    static let shared = HistoryStore()
    private let key = "history"
    private let maxCount = 20

    private(set) var items: [HistoryItem] = []

    private init() {
        load()
    }

    func add(raw: String, polished: String) {
        items.insert(HistoryItem(date: Date(), raw: raw, polished: polished), at: 0)
        if items.count > maxCount {
            items = Array(items.prefix(maxCount))
        }
        save()
    }

    func clear() {
        items = []
        save()
    }

    private func save() {
        let array: [[String: Any]] = items.map {
            ["date": $0.date.timeIntervalSince1970, "raw": $0.raw, "polished": $0.polished]
        }
        UserDefaults.standard.set(array, forKey: key)
    }

    private func load() {
        guard let array = UserDefaults.standard.array(forKey: key) as? [[String: Any]] else { return }
        items = array.compactMap { dict in
            guard let t = dict["date"] as? Double,
                  let raw = dict["raw"] as? String,
                  let polished = dict["polished"] as? String else { return nil }
            return HistoryItem(date: Date(timeIntervalSince1970: t), raw: raw, polished: polished)
        }
    }
}

File: MicTypeWindows/src/MicType.Windows/Core/HistoryStore.cs
using System.IO;
using System.Text.Json;

namespace MicType.Win.Core;

public sealed class HistoryStore
{
    public static HistoryStore Instance { get; } = new();
    private const int MaxCount = 20;
    private readonly List<HistoryItem> _items = [];

    private HistoryStore()
    {
        Load();
    }

    public IReadOnlyList<HistoryItem> Items => _items;

    public void Add(string raw, string polished)
    {
        _items.Insert(0, new HistoryItem(DateTimeOffset.Now, raw, polished));
        if (_items.Count > MaxCount)
        {
            _items.RemoveRange(MaxCount, _items.Count - MaxCount);
        }
        Save();
    }

    public void Clear()
    {
        _items.Clear();
        Save();
    }

    private void Save()
    {
        Directory.CreateDirectory(AppPaths.AppDataDir);
        File.WriteAllText(AppPaths.HistoryPath, JsonSerializer.Serialize(_items, new JsonSerializerOptions
        {
            WriteIndented = true
        }));
    }

    private void Load()
    {
        try
        {
            if (!File.Exists(AppPaths.HistoryPath)) return;
            var items = JsonSerializer.Deserialize<List<HistoryItem>>(File.ReadAllText(AppPaths.HistoryPath));
            if (items is null) return;
            _items.Clear();
            _items.AddRange(items.Take(MaxCount));
        }
        catch (Exception ex)
        {
            Log.Error(ex, "Failed to load history");
            _items.Clear();
        }
    }
}
